Fix ls: panicking on dangling symlink with ``--color=auto -l``

AnirbanHalder654322 opened the pull request:
fixes #6227 

The ``get_metadata()`` function seemed to try to deference to the non existent file. We change it to directly get the metadata of the symlink itself incase the deference fails.

BenWiederhake commented on src/uu/ls/src/ls.rs:
```diff
@@ -3354,8 +3354,7 @@ fn color_name(
         // should not exit with an err, if we are unable to obtain the target_metadata
         let md = get_metadata_with_deref_opt(target.p_buf.as_path(), path.must_dereference)
-            .unwrap_or_else(|_| target.get_metadata(out).unwrap().clone());
-
+            .unwrap_or_else(|_| path.p_buf.symlink_metadata().ok().unwrap().clone());
         apply_style_based_on_metadata(path, Some(&md), ls_colors, style_manager, &name)
```
So I guess the bug was raised during the "unwrap"? In that case I have to ask the question: What do we do if `symlink_metadata` ever fails?

Also, is it even the right thing to do? Afterall, the metadata are to do presentation-stuff (color, hyperlinks, etc), so perhaps this might even be the right way to do it:

```suggestion
        let md = get_metadata_with_deref_opt(target.p_buf.as_path(), path.must_dereference);
        apply_style_based_on_metadata(path, md, ls_colors, style_manager, &name)
```

I don't know which approach is correct, please check against GNU 9.5.

AnirbanHalder654322 replied:
Yeah, so basically what happened is the ``target.get_metadata(out).unwrap().clone())`` called unwrap on a None. The ``get_metadata`` basically tries to get the metadata for the non-existent file in case of dangling symlinks. Panics  only happens we use a path to a file.  

The  ``symlink_metadata`` according to the docs  https://doc.rust-lang.org/std/fs/fn.symlink_metadata.html, only Errs in two cases, when path doesn't exist , which we already seem to handle elsewhere (tested it with a bunch of non-existent files) and the other seems to when we can't call a metadata on the file , I couldn't find a file where that is the case, would definitely appreciate some leads for it, so i can test.

But we can just remove unwraps all together with something like this .
 ```
        let md_res = get_metadata_with_deref_opt(target.p_buf.as_path(), path.must_dereference);
        let md = md_res.or(target.p_buf.symlink_metadata());

        apply_style_based_on_metadata(path, md.as_ref(),ok(), ls_colors, style_manager, &name)
```
We still need to call ``symlink_metadata()`` to color the file being pointed to by the color of the symlink itself.

BenWiederhake replied:
Thanks, this answers all my questions and worries :D

(Technically the file can still be missing due to TOCTOU, but that bug is literally everywhere, so I'll ignore it here.)

BenWiederhake reviewed (changes requested):
Thanks for tackling this!
- [X] I'm not yet entirely convinced that the second invocation is actually needed
- [X] If it is needed, what do we do if it fails?
- [ ] And of course, please write tests so that we can't regress if we change this in the future :D
- [ ] EDIT: Oh, and please fix the typo in the commit message :)

AnirbanHalder654322 commented:
Changes since last push :

- Changed the logic to use       
```
        let md_res = get_metadata_with_deref_opt(target.p_buf.as_path(), path.must_dereference);
        let md = md_res.or(target.p_buf.symlink_metadata());

        apply_style_based_on_metadata(path, md.as_ref(),ok(), ls_colors, style_manager, &name)
```
- Added test

BenWiederhake commented on tests/by-util/test_ls.rs:
```diff
@@ -1376,6 +1376,84 @@ fn test_ls_long_symlink_color() {
+
+    let stdout = result.stdout_str();
+
+    // Helper function to isolate the sequence in the string from where the ANSI C escape code begins
```
Why is parsing required here? Shouldn't it be possible to do something like `.stdout_contains("\x1b[0m\x1b[40;31;01mdir1/dangling_symlink\x1b[0m -> \x1b[40;31;01minvalid\x1b[0m\n")`?
We don't need nice test-failure messages if a color is slightly wrong. Parsing is yet another opportunity to fail, so if we can get away with not parsing anything at all, then I would prefer that.

AnirbanHalder654322 replied:
Yeah this makes sense. I was avoiding regex due to my inexperience with it. I will switch to matching against a regex .

BenWiederhake commented on tests/by-util/test_ls.rs:
```diff
@@ -1376,6 +1376,84 @@ fn test_ls_long_symlink_color() {
+    let at = &ts.fixtures;
+
+    at.mkdir("dir1");
+    ts.ccmd("ln")
```
`at.symlink_file` can do this for you.

BenWiederhake reviewed (changes requested):
Fixes the problem, and adds a test. Yay!

It looks to me like the test can be simplified by a lot, what did I miss?

AnirbanHalder654322 commented:
Changes since last push: 

- Using ``at.symlink_file`` for symlink creation
- Switched to regex matching for extracting colors from stdout
- Squashed the commit into an older comment formatting commit to clean up history .

BenWiederhake reviewed (commented):
Just one last nitpick, then it can be merged :)

AnirbanHalder654322 commented on tests/by-util/test_ls.rs:
```diff
@@ -1381,25 +1381,14 @@ fn test_ls_long_symlink_color() {
-    // The string between the color start and the color end is the file name itself.
-    type Name = String;
-
+    at.mkdir("dir2");
```
None, i did some local testing with foo being in another folder and forgot to remove it :)

AnirbanHalder654322 commented:
I was wondering if i should clean up the commit history a bit more ? Maybe only leave two commits one for the main change and another for the test ?

BenWiederhake commented on tests/by-util/test_ls.rs:
```diff
@@ -1376,6 +1376,43 @@ fn test_ls_long_symlink_color() {
+/// This test is for "ls -l --color=auto|--color=always"
+/// We use "--color=always" as the colors are the same regardless of the color option being "auto" or "always"
+/// using 'ln' to create a dangling symlink
+/// Only tests whether the specific color of the target and the dangling_symlink are equal
```
Hmm, this description doesn't match the code. In particular, the code does *not* check whether the color of the target and the symlink are equal, or in fact whether it matches any color in particular.

Are the colors identical on every platform?
- If yes, the regex could be simplified to just `\x1b[01;36m`.
- If not, either the description should change (to match what the code actually does) or the code should change (e.g. `assert_eq!(colors_vec[0], colors_vec[1]);`)

AnirbanHalder654322 replied:
1. Somebody can change the default colors for the terminal by changing them in ``LS_COLORS`` env variable. I anticipate that will create problems for what the test would show. 
2. I kind of like the fact that we are checking the color code along with the file path. I will add this ``assert_eq!(colors_vec[0], colors_vec[1]);``  and keep the rest of the test as it is and modify to description to match what we are doing.

BenWiederhake reviewed (approved):
LGTM :+1:

BenWiederhake reviewed (approved):
Still looks good ^^
